lib/spack/spack/repository.py

```diff
@@ -112,48 +128,138 @@ def __init__(self, *repo_dirs, **kwargs):
         self._all_package_names = None
         self._provider_index = None
 
-        # If repo_dirs is empty, just use the configuration
-        if not repo_dirs:
-            import spack.config
-            repo_dirs = spack.config.get_config('repos')
-            if not repo_dirs:
-                raise NoRepoConfiguredError(
-                    "Spack configuration contains no package repositories.")
-
         # Add each repo to this path.
         for root in repo_dirs:
-            try:
-                repo = Repo(root, self.super_namespace)
-                self.put_last(repo)
-            except RepoError as e:
-                tty.warn("Failed to initialize repository at '%s'." % root,
-                         e.message,
-                         "To remove the bad repository, run this command:",
-                         "    spack repo rm %s" % root)
+            self.append_from_path(root)
+
+    def append_from_path(self, root):
+        """Append a repository to the list of managed ones. Takes a path
+        as argument.
+
+        Args:
+            root: path to the repository
+        """
+        try:
+            repo = Repo(root, self.super_namespace)
+            self.append(repo)
```

Comment:
> In my opinion inheriting from MutableSequence makes it harder to modify because it's harder to understand how it works.

and

> Speaking generically, enforcing the stronger contract is part of what makes it harder to modify

Speaking generically, that may be true (and that's probably why you are rarely advised to come up with your own abc).

In this case, instead, enforcing a contract that makes a class *behave like a built-in structure of the language* results in code that is easier to modify, because you know what to expect. Besides `swap`, in the current implementation we have:
```python
RepoPath.put_first(x) # Is it like [].insert(0, x) ?
RepoPath.put_last(x)  # Is it like [].append(x) ?
RepoPath.remove(x) # Is it like [].remove(x) ?
RepoPath.first_repo() # Is it the same as [].__getitem__(0) ?
```
With a class that resembles a list sometimes, but does not other times, following your reasoning we'll need to inspect the semantic of every function before modifying one. With a `MutableClass` we don't: as soon as we implement the prescribed abstract functions with the same semantic as list, every other method that is part of the protocol is already implemented and works "as expected", i.e. as a list. 

A counter-example on the fact that it's not so difficult to modify this class (as compared to the previous implementation) is that the different semantic for `__contains__` has been ported just maintaining the override, and making the new, unexpected behavior explicit in the docs.

Another example focused on encapsulation: with the current interface I can construct a list of repos, and then get back only the first element through its interface. To get other elements or iterate over them I need to access "implementation details" of my object, like `spack.repo.repos`. That wouldn't be bad per-se in python, but in this case we have three states that must be kept consistent (`repos`, `by_path`, `by_namespace`), and we are placing this responsibility on clients of `RepoPath`. Using a `MutableSequence` we can operate on the object API directly and modify its state in a consistent way, like if it was a list.

In the end for me this is all about:

- having a consistent interface across the code, for behaviors that are similar (list, dictionaries, tuples, etc.)
- improving encapsulation of the data structure we use

You're right that these behaviors are not yet used in the code, but probably we'll need to touch repos to solve #3115. This modification in my opinion is a step toward making `RepoPath` more flexible to work with, and it doesn't hurt (at least imo).